workers/generic-worker/d2g_feature.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/taskcluster/taskcluster/v83/workers/generic-worker/host"
+)
+
+type (
+	D2GFeature struct {
+	}
+
+	D2GTaskFeature struct {
+		task *TaskRun
+	}
+)
+
+func (df *D2GFeature) Name() string {
+	return "D2G"
+}
+
+func (df *D2GFeature) Initialise() error {
+	return nil
+}
+
+func (df *D2GFeature) IsEnabled() bool {
+	return config.D2GEnabled()
+}
+
+func (df *D2GFeature) IsRequested(task *TaskRun) bool {
+	return task.D2GInfo != nil
+}
+
+func (df *D2GFeature) NewTaskFeature(task *TaskRun) TaskFeature {
+	return &D2GTaskFeature{
+		task: task,
+	}
+}
+
+func (dtf *D2GTaskFeature) ReservedArtifacts() []string {
+	return []string{}
+}
+
+func (dtf *D2GTaskFeature) RequiredScopes() scopes.Required {
+	return scopes.Required{}
+}
+
+func (dtf *D2GTaskFeature) Start() *CommandExecutionError {
+	imageLoader := dtf.task.D2GInfo.Image.ImageLoader()
+	out, err := host.Output(
+		"/usr/bin/env",
+		"bash",
+		"-c",
+		imageLoader.LoadCommand(taskContext.TaskDir),
+	)
+	if err != nil {
+		return executionError(internalError, errored, fmt.Errorf("could not load docker image: %v\n%v", err, out))
+	}
+
+	if dtf.task.Payload.Env == nil {
+		dtf.task.Payload.Env = make(map[string]string)
+	}
+
+	dtf.task.Payload.Env["D2G_IMAGE_ID"] = strings.TrimSpace(out)
+	err = dtf.task.setVariable("D2G_IMAGE_ID", strings.TrimSpace(out))
+	if err != nil {
+		return executionError(internalError, errored, fmt.Errorf("could not set D2G_IMAGE_ID environment variable: %v", err))
+	}
+
+	if dtf.task.DockerWorkerPayload.Features.ChainOfTrust {
+		out, err := host.Output(
+			"/usr/bin/env",
+			fmt.Sprintf("D2G_IMAGE_ID=%s", strings.TrimSpace(out)),
+			"bash",
+			"-c",
+			imageLoader.ChainOfTrustCommand(taskContext.TaskDir),
+		)
+		if err != nil {
+			return executionError(internalError, errored, fmt.Errorf("could not create chain of trust additional data file: %v\n%v", err, out))
+		}
+	}
+	return nil
+}
+
+func (dtf *D2GTaskFeature) Stop(err *ExecutionErrors) {
+	if dtf.task.DockerWorkerPayload.Features.DockerSave {
+		out, e := host.Output(
+			"docker",
+			"commit",
+			dtf.task.D2GInfo.ContainerName,
+			dtf.task.D2GInfo.ContainerName,
+		)
+		if e != nil {
+			err.add(executionError(internalError, errored, fmt.Errorf("could not commit docker container: %v\n%v", e, out)))
+		}
+		out, e = host.Output(
+			"/usr/bin/env",
+			"bash",
+			"-c",
+			fmt.Sprintf("docker save %s | gzip > image.tar.gz", dtf.task.D2GInfo.ContainerName),
+		)
+		if e != nil {
+			err.add(executionError(internalError, errored, fmt.Errorf("could not save docker image: %v\n%v", e, out)))
+		}
+	}
+
+	for _, artifact := range dtf.task.D2GInfo.CopyArtifacts {
+		out, e := host.Output(
+			"docker",
+			"cp",
+			fmt.Sprintf("%s:%s", dtf.task.D2GInfo.ContainerName, artifact.SrcPath),
+			filepath.Join(taskContext.TaskDir, artifact.DestPath),
+		)
+		if e != nil {
+			dtf.task.Warnf("Artifact %q not found at %q: %v\n%v", artifact.Name, artifact.SrcPath, e, out)
```

Comment:
Just log a warning if the artifact isn't found in the container, as docker worker does: https://github.com/taskcluster/taskcluster/blob/main/workers/docker-worker/src/features/artifacts.js#L52-L55